### DISCUSSION TITLE
How to use the `SendInput` method correctly?

### DISCUSSION THREAD

**Discussion by vitkuz573** [microsoft/CsWin32#817] [Q&A] [answered]
I'm trying to use the `SendInput` method and don't quite understand how to do it right. At the moment my code looks something like this (example):

```csharp
if (deltaY < 0)
{
    deltaY = -120;
}
else if (deltaY > 0)
{
    deltaY = 120;
}

var union = new INPUT._Anonymous_e__Union
{
    mi = new MOUSEINPUT
    {
        dwFlags = MOUSE_EVENT_FLAGS.MOUSEEVENTF_WHEEL,
        dx = 0,
        dy = 0,
        time = 0,
        mouseData = deltaY,
        dwExtraInfo = 0
    }
};

var input = new INPUT
{
    type = INPUT_TYPE.INPUT_MOUSE,
    Anonymous = union
};

var inputs = new Span<INPUT>();
inputs.Fill(input);

SendInput(inputs, Marshal.SizeOf(typeof(INPUT)));
```

But it doesn't work. That's why I have this question.

**Comment by AArnott** [comment, marked as answer]
Ultimately I think your error is that you create a span of zero length, and then try to 'fill' it, but nothing will happen because it's empty.

```diff
-var inputs = new Span<INPUT>();
-inputs.Fill(input);
+var inputs = new Span<INPUT>(ref input);
```

But that relies on a new constructor for .NET 7. If you're not that lucky, this will also work:

```cs
Span<INPUT> inputs = stackalloc INPUT[1];
inputs[0] = input;
```

I would also touch up a bit of the other syntax, leaving me with:

```cs
var input = new INPUT
{
    type = INPUT_TYPE.INPUT_MOUSE,
    Anonymous =
    {
        mi = new MOUSEINPUT
        {
            dwFlags = MOUSE_EVENT_FLAGS.MOUSEEVENTF_WHEEL,
            dx = 0,
            dy = 0,
            time = 0,
            mouseData = deltaY switch { < 0 => -120, > 0 => 120, 0 => 0 },
            dwExtraInfo = 0
        },
    },
};

#if NET7_0_OR_GREATER
Span<INPUT> inputs = new(ref input);
#else
Span<INPUT> inputs = stackalloc INPUT[1];
inputs[0] = input;
#endif

PInvoke.SendInput(inputs, Marshal.SizeOf(typeof(INPUT)));
```

I haven't tested this, but I'm optimistic it'll work. :)

**Reply by vitkuz573** [reply]
I checked now, it doesn't work that way. Not only for `MouseWheel`, it also does not work for `MouseMove`.

**Reply by AArnott** [reply]
Can you elaborate on "it doesn't work"? Do you get an error back?

**Reply by vitkuz573** [reply]
When I output the value `var result = SendInput(inputs, Marshal.SizeOf(typeof(INPUT)));` I see `0` and the code is not executed.

**Reply by vitkuz573** [reply]
Sorry, in another use case everything works. Apparently I'm missing something, I'll look into it. Thanks for the reply!

**Reply by Zintom** [reply]
How do you supply a negative value to `mouseData`?, it is an unsigned integer.

**Reply by vitkuz573** [reply]
It worked for me

```csharp
mouseData = data.DeltaY < 0 ? 120u : data.DeltaY > 0 ? unchecked((uint)-120) : 0u
```

**Reply by AArnott** [reply]
@vitkuz573's workaround is correct, and unfortunately this isn't something we can look forward to being improved in the metadata. https://github.com/microsoft/win32metadata/issues/1865#issuecomment-1977365435

**Reply by Zintom** [reply]
Thanks, I resolved it by casting the result of the switch expression as a whole which worked too.
